### PR TITLE
Plane: added Q_ASSIST_ALT2

### PR DESCRIPTION
**Pull request by tridge**
This adds Q_ASSIST_ALT2 and Q_ASSIST_ALT_ACT. This gives more complex altitude assistance behaviour:

 - the altitude specified in Q_ASSIST_ALT2 is latching, so assistance is only enabled once above that height by 2 meters
 - if the assistance altitude is breached then the Q_ASSIST_ALT_ACT action is used to control the action. Choices are: None, ClimbGuided, ClimbRTL, QHover, QLoiter, QLand, QRTL
  - for the climb actions the aircraft will climb 35m in GUIDED mode

This is added at the request of Alex from AraceUAS

### CODE REVIEW

File: ArduPlane/quadplane.h [hunk 1]
```diff
@@ -298,10 +299,27 @@ class QuadPlane
...
+    struct {
+        // altitude to trigger assistance
+        AP_Int16 alt;
+        AP_Int16 alt2;
+        AP_Enum<AltAssistAction> action;
+        uint32_t start_ms;
+        bool in_assist;
+        bool alt2_enabled;
+        bool reached_climb_alt;
+        const float pre_climb_alt = 35;
```
Comment (IamPete1) [inline ArduPlane/quadplane.h:321]: Should this be a define? it seems a little hidden down here.

File: ArduPlane/quadplane.cpp [hunk 2]
```diff
@@ -1330,26 +1388,51 @@ bool QuadPlane::should_assist(float aspeed, bool have_airspeed)
...
+    if (should_assist_alt && assist_alt.alt > 0) {
+        float height_above_ground = plane.relative_ground_altitude(plane.g.rangefinder_landing);
+        if (height_above_ground < assist_alt.alt) {
+            if (assist_alt.start_ms == 0) {
+                assist_alt.start_ms = now;
+            }
+            if (now - assist_alt.start_ms > assist_delay*1000) {
+                // we've been below assistant alt for Q_ASSIST_DELAY seconds
+                if (!assist_alt.in_assist) {
+                    start_alt_assist(height_above_ground);
```
Comment (IamPete1) [inline ArduPlane/quadplane.cpp:1405]: This seems to be doing the assistance action based on the first assistance alt not the new one?

File: ArduPlane/quadplane.cpp [hunk 3]
```diff
@@ -1330,26 +1388,51 @@ bool QuadPlane::should_assist(float aspeed, bool have_airspeed)
...
+        }
+    }
+    if (should_assist_alt && assist_alt.alt2 > 0) {
         float height_above_ground = plane.relative_ground_altitude(plane.g.rangefinder_landing);
-        if (height_above_ground < assist_alt) {
-            if (alt_error_start_ms == 0) {
-                alt_error_start_ms = now;
+        if (!assist_alt.alt2_enabled) {
+            if (height_above_ground > assist_alt.alt2+2) {
+                assist_alt.alt2_enabled = true;
```
Comment (IamPete1) [inline ArduPlane/quadplane.cpp:1418]: We should probably reset this on landing / disarm

File: ArduPlane/quadplane.cpp [hunk 4]
```diff
@@ -1283,6 +1298,47 @@ float QuadPlane::desired_auto_yaw_rate_cds(void) const
...
+            plane.set_mode(plane.mode_qland, ModeReason::ALT_ASSIST);
+        } else {
+            plane.set_mode(plane.mode_qloiter, ModeReason::ALT_ASSIST);
+        }
+        break;
+    case AltAssistAction::QLand:
+        plane.set_mode(plane.mode_qland, ModeReason::ALT_ASSIST);
+        break;
+    case AltAssistAction::QRTL:
+        plane.set_mode(plane.mode_qrtl, ModeReason::ALT_ASSIST);
```
Comment (IamPete1) [inline ArduPlane/quadplane.cpp:1329]: I wonder if we need to make sure this is actually a QRTL, and does not switch to forward flight and do an approach. My understanding is that if this assist alt ever trips is suggests that the vehicle may not be capable of forward flight at all.

**Comment by IamPete1** [comment]
I wonder if we should disable for manual forward flight modes.

**Comment by pompecukor** [comment]
There needs to be a logic to prevent it from triggering on false positives. That is on commanded/intentional altitude changes in autonomous modes. For instance a loitering to alt jut before transition on landing.
Or even if the new waypoint is lower AGL than the trigger alt.

**Comment by tridge** [comment]
> 

we can certainly cover the major cases, but some cases will be tricky.
What I first need to know is what cases should be considered as false positives:
 - is CRUISE or FBWB mode, with the pilot commanding a descent using the sticks, considered a false positive, or should the assistance trigger?
 - what about LOITER mode, with the new option to change alt with pitch stick?
 - what about QLOITER and QHOVER? Normally we wouldn't do Q_ASSIST in those modes as we're already in a VTOL mode, so VTOL assistance doesn't make sense.

The core use case will be AUTO mode and how to handle that. Perhaps it should only consider triggering if the TECS height error shows it is under the current TECS target altitude by a margin? That will still risk triggering if either the TECS tune is not good enough, or we get a temporary local terrain change that pushes us past the margin.
If we only look at the target WP altitude then we would have two problems:
 - at the point we start on a waypoint we may not have the terrain data for the WP in ram cache, so we can't lookup the terrain height of the target WP. That could cause a false trigger
 - the target WP may be above the assist alt2 limit, but we are currently under that limit, for example, we're transitioning from a mode like GUIDED where a low alt has been commanded

We also need to work out how this relates to the geofence.

**Comment by pompecukor** [comment]
@tridge All very excellent points.

I would agree that it should be completely off any of Qmodes. That stands to reason.

Also I would think the main relevance is in Auto mode (including guided, loiter and RTL).
For semi auto and manual modes I would not have any reason to want it to ignore the trigger. That is for two reasons.
1. In such cases operator would be aware of the settings so should not intentionally go below said altitude.
2. If they end up doing so, they can always switch modes back to FBWA/B or whatever. It only triggers once as far as I understand.

I would not go with the TECS route. Adding an immeasurable amount of complexity. That almost defeats the propose.

I think it is fair sacrifice for it not to be able to know terrain data at the start. Please remember the only sensible scenario for there to be a WP that is lower than this trigger, would be if setting up an manually plotted approach for landing. That is not and auto loiter, but rather the old school step down AGL from one WP to the other as we approach home and about to transition.
This would be one of the most common cases of false trigger.
The other as I mentioned is with regards to landing, when using QRTL 2 or even mission based Loiter to alt to decrease the altitude prior to transition. Same goes for the new Loiter_to_Qland mode.

As far as we are concerned, if those as sorted (that is any alt that is commanded in mission or autonomy that goes below the Alt assist2 is ignored), then we are happy to use it. If you do add any extra stuff, then obviously we wont resist.

So to summarize, I think it would be usable and acceptable as a base line at this time (remembering that it is an option and not mandatory to activate). 

- Have it fully active and trigger in and none and semi-autonomous modes (FBWA/B, Manual, etc). No risk concern about false positives. User is controlling and can switch modes back to whatever they desire.
- Have it completely deactivated in all Qmodes.
- Prevent activation in Autonomous modes (auto, loiter, guided) when the upcoming waypoint is lower than assist alt2.

Triggers would use terrain data if available. 

With the above the PR will serve it primary purpose and be usefully/usable already (can be tested and merged), then you can fine tune in spare time as needed.

**Comment by pompecukor** [comment]
> I wonder if we should disable for manual forward flight modes.

It is actually very useful and manual forward flight modes, especially during setup and tuning. Adds extra safety. We used this a lot. Remember you set the desired altitude.

**Comment by pompecukor** [comment]
@tridge and chance on follow up on this one?

**Comment by IamPete1** [comment]
I spent some time poking about at this. I have come to the conclusion that it should be a min altitude fence. We would just need to add QRTL and QLAND fence actions. Possibly some capability to do a different action for min altitude vs polygon.
